File: login_facebook.js
/**
 * Run ONCE to save your Facebook session.
 * Usage: npx playwright@latest node login_facebook.js
 */
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

const SESSION_PATH = path.join(__dirname, 'orchestrator', 'sessions', 'facebook_session.json');

async function main() {
  fs.mkdirSync(path.dirname(SESSION_PATH), { recursive: true });

  console.log('\n=== Facebook Login ===');
  console.log('A browser window will open.');
  console.log('Log into Facebook normally.');
  console.log('The window closes automatically once you are logged in.\n');

  const browser = await chromium.launch({ headless: false, slowMo: 50 });
  const context = await browser.newContext({
    viewport: { width: 1280, height: 800 },
    userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
  });

  const page = await context.newPage();
  await page.goto('https://www.facebook.com/login');

  console.log('Waiting for you to log in...');

  try {
    await page.waitForURL('https://www.facebook.com/', { timeout: 180000 });
  } catch {
    try {
      await page.waitForFunction(
        () => window.location.hostname === 'www.facebook.com' && !window.location.pathname.includes('login'),
        { timeout: 180000 }
      );
    } catch {
      console.log('Timed out. Try running again.');
      await browser.close();
      return;
    }
  }

  const storage = await context.storageState();
  fs.writeFileSync(SESSION_PATH, JSON.stringify(storage, null, 2));
  await browser.close();

  console.log('\n✅ Facebook session saved!');
  console.log(`   Location: ${SESSION_PATH}`);
  console.log('\nAgents can now send Facebook DMs as you.\n');
}

main().catch(console.error);

File: login_facebook.py
"""
Run this ONCE to save your Facebook session.
After this, agents can send DMs as you forever.

Usage:
  python login_facebook.py

Requirements:
  pip install playwright
  playwright install chromium
"""
import asyncio
import json
from pathlib import Path

async def main():
    try:
        from playwright.async_api import async_playwright
    except ImportError:
        print("Installing playwright...")
        import subprocess
        subprocess.run(["pip", "install", "playwright"], check=True)
        subprocess.run(["playwright", "install", "chromium"], check=True)
        from playwright.async_api import async_playwright

    SESSION_PATH = Path(__file__).parent / "orchestrator" / "sessions" / "facebook_session.json"
    SESSION_PATH.parent.mkdir(parents=True, exist_ok=True)

    print("\n=== Facebook Login ===")
    print("A browser window will open.")
    print("Log into Facebook normally.")
    print("The window will close automatically once you're logged in.\n")

    async with async_playwright() as pw:
        browser = await pw.chromium.launch(headless=False, slow_mo=50)
        context = await browser.new_context(
            viewport={"width": 1280, "height": 800},
            user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
        )
        page = await context.new_page()
        await page.goto("https://www.facebook.com/login")

        print("Waiting for you to log in...")

        try:
            # Wait up to 3 minutes for login to complete
            await page.wait_for_url("https://www.facebook.com/", timeout=180000)
        except Exception:
            try:
                # Sometimes lands on a different URL after login
                await page.wait_for_function(
                    "() => window.location.href.includes('facebook.com') && !window.location.href.includes('login')",
                    timeout=180000
                )
            except Exception:
                print("Timeout — did you log in? Try running again.")
                await browser.close()
                return

        # Save session
        storage = await context.storage_state()
        with open(SESSION_PATH, "w") as f:
            json.dump(storage, f)

        await browser.close()

    print(f"\n✅ Facebook session saved!")
    print(f"   Location: {SESSION_PATH}")
    print(f"\nAgents can now send Facebook DMs as you.")
    print("You never need to do this again unless you log out of Facebook.\n")

if __name__ == "__main__":
    asyncio.run(main())
